Valkey master-replica support

olehoerb opened the pull request:
## Description
feat: Implements complete Valkey master-replica support with leader election-based backup-restore coordination. Ensures safe backup and restore operations in multi-replica deployment.
Only Valkey-master (pod-0) performs backup/restore operations

- Add leader election package using https://pkg.go.dev/k8s.io/client-go/tools/leaderelection for coordination
- Add Valkey master-replica database implementation
- Add integration tests for master-replica restore workflow
- Add Kubernetes deployment examples and backing resources
- Integrate leader election checks into backup coordination

References: #124
To close #124 helm chart is still needed

olehoerb committed:
valkey: added cluster mode support.
-cluster setup not working correctly

olehoerb committed:
added simple leader election functionality as described in issue #124

olehoerb committed:
wip: fix race condition in leader election
- fix cluster logic issues

olehoerb committed:
fix: fixes issues for production readiness
- fix backup leader/master mismatch
- Add context management
- Optimize backup coordination

olehoerb committed:
fix: verify database role in leader election for valkey master-replica

refactored the name from valkey-cluster to valkey-master-replica

olehoerb committed:
add valkey-master-replica-local.yaml

olehoerb committed:
fix: remove ObjectMeta from embedded field selectors

majst01 commented on cmd/internal/database/valkey/valkey.go:
```diff
@@ -0,0 +1,359 @@
+		return -1
+	}
+
+	ordinalStr := parts[len(parts)-1]
```
consider strings.Cut

olehoerb replied:
Thanks for pointing it out! But I think I found an even better solution here after reviewing the strings functions

majst01 commented on deploy/valkey-local.yaml:
```diff
@@ -132,20 +155,95 @@ status:
+    \ exec valkey-server --port 6379 --bind 0.0.0.0\nelse\n  echo \"I am a replica
+    (pod-$ORDINAL), connecting to master at valkey-0.valkey.${POD_NAMESPACE}.svc.cluster.local\"\n
+    \ exec valkey-server --port 6379 --bind 0.0.0.0 --replicaof valkey-0.valkey.${POD_NAMESPACE}.svc.cluster.local
+    6379\nfi\n"
```
Can be done with a multiline yaml content e.g. |

majst01 commented on deploy/valkey-master-replica-local.yaml:
```diff
@@ -0,0 +1,309 @@
+    \ exec valkey-server --port 6379 --bind 0.0.0.0\nelse\n  echo \"I am a replica
+    (pod-$ORDINAL), connecting to master at valkey-0.valkey.${POD_NAMESPACE}.svc.cluster.local\"\n
+    \ exec valkey-server --port 6379 --bind 0.0.0.0 --replicaof valkey-0.valkey.${POD_NAMESPACE}.svc.cluster.local
+    6379\nfi\n"
```
multiline content

majst01 reviewed (changes requested):
Nice work, first simple comments from my side

olehoerb committed:
fix: use lastIndex instead of Split and multiline YAML format for configMaps

Gerrit91 commented on cmd/internal/backup/backup.go:
```diff
@@ -59,6 +59,14 @@ func (b *Backuper) Start(ctx context.Context) error {
+				b.log.Debug("skipping backup - not elected as leader")
+				return
+			}
+		}
```
Better use an extended contract in the `DatabaseProber` interface instead of a type cast, something like:

```
	if !b.db.IsLeader(ctx) {
			b.log.Debug("skipping backup - not elected as leader")
			return
		}
```

Gerrit91 commented on Makefile:
```diff
@@ -46,6 +46,12 @@ test-integration-valkey: kind-cluster-create
+.PHONY: test-integration-valkey-master-replica
+test-integration-valkey-master-replica: kind-cluster-create
+	kind --name backup-restore-sidecar load docker-image ghcr.io/metal-stack/backup-restore-sidecar:latest
+	kind --name backup-restore-sidecar load docker-image ghcr.io/valkey-io/valkey:8.1-alpine
```
Can be removed as this comes from the internet.

Gerrit91 commented on cmd/internal/database/valkey/valkey.go:
```diff
@@ -0,0 +1,362 @@
+	}
+	if podNamespace == "" {
+		return nil, fmt.Errorf("cluster mode requires POD_NAMESPACE environment variable to be set")
+	}
```
These checks can be moved into the leader election package if they are required for it.

Gerrit91 commented on cmd/internal/database/valkey/valkey.go:
```diff
@@ -0,0 +1,362 @@
+	client *redis.Client
+
+	clusterMode     bool
+	clusterSize     int
```
This field is not used and can probably be removed.

Gerrit91 commented on pkg/generate/examples/examples/valkey.go:
```diff
@@ -232,18 +355,202 @@ func ValkeyBackingResources(namespace string) []client.Object {
 backup-cron-schedule: "*/1 * * * *"
-object-prefix: valkey-test
+object-prefix: valkey-test-${POD_NAME}
 redis-addr: localhost:6379
```
This key is not used for the valkey backend, so it can be removed.

Gerrit91 commented on pkg/generate/examples/examples/valkey.go:
```diff
@@ -232,18 +355,202 @@ func ValkeyBackingResources(namespace string) []client.Object {
+backup-provider: local
+backup-cron-schedule: "*/1 * * * *"
+object-prefix: valkey-master-replica-test-${POD_NAME}
+redis-addr: localhost:6379
```
This key is not used for the valkey backend, so it can be removed.

Gerrit91 commented on cmd/internal/database/valkey/valkey.go:
```diff
@@ -0,0 +1,362 @@
+	log.Info("Creating Valkey instance", "clusterMode", clusterMode, "clusterSize", clusterSize)
+
+	v.client = redis.NewClient(&redis.Options{
+		Addr:     "localhost:6379",
```
This should ideally come from a configuration parameter (e.g. `"valkey-addr"`)

Gerrit91 commented on cmd/internal/database/valkey/valkey.go:
```diff
@@ -0,0 +1,362 @@
+	}
+
+	if !isMaster {
+		db.log.Info("elected as backup leader but not Valkey master, skipping backup")
```
How do you ensure that the leader election changes to the database leader to prevent there are no backups taken in case of a constant mismatch?

Gerrit91 commented on cmd/internal/database/valkey/valkey.go:
```diff
@@ -0,0 +1,362 @@
+	v := &Valkey{
+		log:             log,
+		datadir:         datadir,
+		password:        getPassword(password),
```
You can replace the `getPassword` function in favor of a function from [metal-lib](https://github.com/metal-stack/metal-lib/blob/master/pkg/pointer/pointer.go):

```suggestion
		password:         pointer.SafeDerefOrDefault(password, ""),
```

Gerrit91 commented on cmd/internal/database/valkey/valkey.go:
```diff
@@ -0,0 +1,362 @@
+
+					// Leader election considers database role: only restore if this pod will be the Valkey master
+					// In master-replica mode, pod-0 is always the master (determined by init.sh)
+					podName := os.Getenv("POD_NAME")
```
Maybe it's more secure to retrieve this from the `leaderElection` package because the pod name might originate from the configuration and not only from the environment variable.

Gerrit91 commented on cmd/internal/database/valkey/valkey.go:
```diff
@@ -0,0 +1,362 @@
+		return fmt.Errorf("restore file not present: %s", dump)
+	}
+
+	if db.clusterMode {
```
Probably it's better to reduce code intention by negating this expression, the `for` loop in this function has explicit returns anyway so last line of this function does not need to be repeated.


```suggestion
	if !db.clusterMode {
```

Gerrit91 reviewed (commented):
Actually looks really good and promising!

Not fully done yet but here are some first review comments.

Gerrit91 commented on pkg/generate/examples/examples/valkey.go:
```diff
@@ -43,7 +45,8 @@ func ValkeySts(namespace string) *appsv1.StatefulSet {
 				Spec: corev1.PodSpec{
-					HostNetwork: true,
+					HostNetwork:        true,
+					ServiceAccountName: "valkey-backup-restore",
```
It would probably be good to add a topology spread constraint here for an example on the node name, like:

```yaml
      topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            app: valkey
        maxSkew: 1
        topologyKey: kubernetes.io/hostname
        whenUnsatisfiable: ScheduleAnyway
```

Gerrit91 reviewed (commented):
I was able to play around a bit with the setup. Is it correct that this is not a real Valkey Cluster as described [here](https://valkey.io/topics/cluster-tutorial/) but rather just adding replicas, which cannot accept writes and are not promoted to master instances under any circumstances?

I came to this conclusion because I run:

```
❯ k exec -it valkey-master-replica-0 -c valkey -- valkey-cli cluster info
ERR This instance has cluster support disabled
```

When killing `pod-0` constantly, I can still read values from database, but it's not possible to write anymore:

```
❯ k exec -it statefulsets/valkey-master-replica -c valkey -- valkey-cli set foo foo
(error) READONLY You can't write against a read only replica.
```

I think, this is also a good improvement in general as it allows serving read-requests during node roll or whatever. But I am not sure if leader election is really required in this setup because backups and restores can only be done from `pod-0` anyway? Also we should not call it `clusterMode`? 

Can I ask you to somewhere describe the approach in a markdown document in the `docs` folder? I think this would help a lot. :)

olehoerb committed:
fix:
-use isMaster for backup coordination instead of leader election
-fix STATEFUL_NAME mismatch on valkey container
-restore HostNetwork for standalone, disable for master-replica
-fix newValkeyClient retry to actually ping
-added docs describing the approach

olehoerb committed:
fix:
-replaced pointer usage with GO 1.26 pointer

olehoerb committed:
fix:
-replaced pointer usage with GO 1.26 pointer

olehoerb committed:
Merge branch 'master' into valkey-cluster-config

Gerrit91 commented on docs/valkey_master_replica.md:
```diff
@@ -0,0 +1,205 @@
+At restore time, Valkey is not running yet, so the sidecar cannot query the database role. Instead it uses Kubernetes leader election to coordinate which pod performs the restore:
+
+1. All pods compete for a `Lease` resource.
+2. The winner checks its pod ordinal. Only pod-0 (the future master) actually restores from backup.
```
This I do not understand. Why isn't checking the ordinal enough if the backup needs to be restored? What benefit does the leader election bring?

Gerrit91 replied:
Imagine the the scenario all pods are terminated and all volumes are lost. Then the pods start up fresh and the replica wins the election and restores the data. When the master comes up, it will not restore the data because it's not the leader and start without data. Wouldn't the replicas then sync empty data from the master?

olehoerb replied:
Thats actually true. The ordinal check is enough and using ordinal check should fix the issue with syncing

Gerrit91 commented on docs/valkey_master_replica.md:
```diff
@@ -0,0 +1,205 @@
+
+## Topology
+
+This is **not** a Valkey Cluster (which requires `cluster-enabled yes` and has built-in sharding/failover). Instead it is a simple master-replica setup using a Kubernetes StatefulSet:
```
Did you evaluate the cluster mode once? What were the issues why it cannot be used as in general this would be really nice if a node roll of a K8s cluster would not cause any interruptions of the service in terms of write operations.

olehoerb committed:
change leader election with simple ordinal check for restore coordination

olehoerb committed:
fix: never take valkey backups from a replica, keep local dump as recovery point

- Backup() now verifies the database role itself: scheduled backups were already filtered through ShouldPerformBackup, but a manual create-backup via the gRPC API on a replica bypassed the check and would have uploaded the replica's state. Now fails with an explicit error
- retain dump.rdb in the data directory after a backup: together with valkey's automatic snapshotting it is the local recovery point, so a pod restarting with an intact volume resumes from local data and a restore from backup only happens on actual volume loss
- drop the valkey-statefulset-name config option, the pod ordinal is derived from POD_NAME
- fix stale leader election log wording, document accurate failure behavior and recovery points
